src/hotspot/share/logging/logStream.hpp

```diff
@@ -53,60 +59,203 @@ class LogStream : public outputStream {
     void append(const char* s, size_t len);
     void reset();
   };
-  LineBuffer  _current_line;
-  LogTargetHandle _log_handle;
 
-  // Prevent operator new for LogStream.
-  static void* operator new (size_t);
-  static void* operator new[] (size_t);
+private:
+  LineBuffer _current_line;
+protected:
+  BackingLog _backing_log;
+public:
+  explicit LogStreamImpl(BackingLog bl)
+    : _backing_log(bl) {};
+
+  virtual ~LogStreamImpl() {
+    if (_current_line.is_empty() == false) {
+      _backing_log.print("%s", _current_line.buffer());
+      _current_line.reset();
+    }
+  }
+
+  bool is_enabled() {
+    return _backing_log.is_enabled();
+  }
+
+  void write(const char* s, size_t len) override {
+    if (len > 0 && s[len - 1] == '\n') {
+      _current_line.append(s, len - 1); // omit the newline.
+      _backing_log.print("%s", _current_line.buffer());
+      _current_line.reset();
+    } else {
+      _current_line.append(s, len);
+    }
+    update_position(s, len);
+  }
+};
+
+template<typename T>
+LogStreamImpl<T>::LineBuffer::LineBuffer()
+ : _buf(_smallbuf), _cap(sizeof(_smallbuf)), _pos(0)
+{
+  _buf[0] = '\0';
+}
+
+template<typename T>
+LogStreamImpl<T>::LineBuffer::~LineBuffer() {
+  assert(_pos == 0, "still outstanding bytes in the line buffer");
+  if (_buf != _smallbuf) {
+    os::free(_buf);
+  }
+}
+
+// try_ensure_cap tries to enlarge the capacity of the internal buffer
+// to the given atleast value. May fail if either OOM happens or atleast
+// is larger than a reasonable max of 1 M. Caller must not assume
+// capacity without checking.
+template<typename T>
+void LogStreamImpl<T>::LineBuffer::try_ensure_cap(size_t atleast) {
+  assert(_cap >= sizeof(_smallbuf), "sanity");
+  if (_cap < atleast) {
+    // Cap out at a reasonable max to prevent runaway leaks.
+    const size_t reasonable_max = 1 * M;
+    assert(_cap <= reasonable_max, "sanity");
+    if (_cap == reasonable_max) {
+      return;
+    }
+
+    const size_t additional_expansion = 256;
+    size_t newcap = align_up(atleast + additional_expansion, additional_expansion);
+    if (newcap > reasonable_max) {
+      log_info(logging)("Suspiciously long log line: \"%.100s%s",
```

Comment:
Can we safely log from inside logging code?

Reply:
Yes. This call never occurs when the backing log is logged to, so there's no risk of double locking.
This code has also been in the source tree since 2016, so experimentally it seems OK :-). My change moves this inside of `logStream.hpp` and deletes `logStream.cpp` as `LogStreamImpl` is templated.